extensions/ql-vscode/src/variant-analysis/ql.ts

```diff
@@ -0,0 +1,87 @@
+import { dirname } from "path";
+import { findCommonParentDir } from "../common/files";
+import { findPackRoot } from "../common/ql";
+
+/**
+ * This function finds the root directory of the QL pack that contains the provided query files.
+ * It handles several cases:
+ * - If no query files are provided, it throws an error.
+ * - If all query files are in the same QL pack, it returns the root directory of that pack.
+ * - If the query files are in different QL packs, it throws an error.
+ * - If some query files are in a QL pack and some aren't, it throws an error.
+ * - If none of the query files are in a QL pack, it returns the common parent directory of the query files. However,
+ *   if there are more than one query files and they're not in the same workspace folder, it throws an error.
+ *
+ * @param queryFiles - An array of file paths for the query files.
+ * @param workspaceFolders - An array of workspace folder paths.
+ * @returns The root directory of the QL pack that contains the query files, or the common parent directory of the query files.
+ */
+export async function findVariantAnalysisQlPackRoot(
+  queryFiles: string[],
+  workspaceFolders: string[],
+): Promise<string> {
+  if (queryFiles.length === 0) {
+    throw Error("No query files provided");
+  }
+
+  // Calculate the pack root for each query file
+  const packRoots: Array<string | undefined> = [];
+  for (const queryFile of queryFiles) {
+    const packRoot = await findPackRoot(queryFile);
+    packRoots.push(packRoot);
+  }
+
+  if (queryFiles.length === 1) {
+    return packRoots[0] ?? dirname(queryFiles[0]);
+  }
+
+  const uniquePackRoots = Array.from(new Set(packRoots));
+
+  if (uniquePackRoots.length > 1) {
+    if (uniquePackRoots.includes(undefined)) {
+      throw Error("Some queries are in a pack and some aren't");
+    } else {
+      throw Error("Some queries are in different packs");
+    }
+  }
+
+  if (uniquePackRoots[0] === undefined) {
+    return findQlPackRootForQueriesWithNoPack(queryFiles, workspaceFolders);
+  } else {
+    // All in the same pack, return that pack's root
+    return uniquePackRoots[0];
+  }
+}
+
+/**
+ * For queries that are not in a pack, a potential pack root is the
+ * common parent dir of all the queries. However, we only want to
+ * return this if all the queries are in the same workspace folder.
```

Comment:
Can you explain in the code why we want them to be in the same workspace folder

Reply:
It seems wrong to allow queries to be in different workspace folders 🤔 [This](https://github.com/github/vscode-codeql/pull/3289/files#diff-95b4bdfe0ae067210693ab52616487856ee8aa4c3c22f171aea5daebd43e76f7R79) is a the bit where we force that.

I guess we don't strictly need that restriction though, so I can remove it.

Reply:
Sorry that wasn't hugely clear. What I was asking for is to document the intent of the restriction in the code. I think a future developer could work through the code and find the place where we enforce the restriction, but what's useful to document as a comment is why we do this. What benefits does it give and what is it protecting us from?

To be honest, I'm 95% sure it's not a security issue to allow the common parent directory to be outside of the workspace. We don't access or write to any files in that actual parent directory. We construct a new query pack and copy the queries into it using the relative path from the shared parent directory.

I also don't think it will block anybody if we impose this restriction, and I think it's an understandable restriction that people would accept. So I'm not against us applying this restriction.

It's also something we can easily change in the future. It should just be documented so that if we do revisit this decision in the future then we can be more confident we know what we're doing.

Reply:
Ah gotcha. Done!